Awkward v2 transition

lgray opened the pull request:
- [ ] Get all tests working again regardless of laziness in nanoevents (I should break this out into sub-bullets)
- [x] base schema passes tests
- [x] nanoaod schema passes tests
- [x] protodune schema passes tests
- [x] physlite schema passes tests
- [x] delphes schema passes tests
- [x] treemaker schema passes tests

- [x] local executor tests pass
- [x] dask executor passes tests
- [x] parsl executor passes tests
- [ ] spark executor passes tests (requires spark upgrade and new data ingestion step!)

- [x] lookup tools passes tests
- [x] jetmet tools - individual correctors
- [x] jetmet tools - CorrectedJetsFactory (heavily lazy)


- [ ] the above, again, except with laziness implemented in dask_awkward

lgray committed:
start on updating to awkward v2 - this will require a few phases and need a transition to dask-awkward as well

uproot5 RCs as well

lgray committed:
remove deprecated coffea.hist

lgray committed:
fix Weight, etc. pending https://github.com/scikit-hep/awkward/pull/1841

lgray committed:
lumi tools and btagging tests are back

get fixes in awkward/uproot

min python is 3.7 now

3.11 everywhere.

fix up deps (sorry spark)

bring back extras later

3.11 has problems

lgray committed:
fixes for clib wrapper, analysis tools (weights)

lgray committed:
remove old btag csv parser, slow and unmaintained

lgray committed:
jec lookup tools fixes

lgray committed:
restore all lookup tools tests that to not require nanoevents

stash interim changes

flake8

lgray committed:
ressurect more tests, awkward is dropping behaviors in some cases

more tests running but needs fixes in awkward as well

preloaded works now as well

enforce awkward2rc3

lgray committed:
stash wip for dask_awkward work

update pins

black

turn off noisy debugging info

lgray committed:
use/update version pinnings in ci

ci updates

update rc pins to consistent versions

remove overspecification of gh-actions

ditto

lgray committed:
_nplikes is now a private interface, need to move to backend instead

lgray committed:
auto schema passes tests

lgray committed:
rename auto schema file to something more obvious

lgray committed:
pdune schema passes tests

lgray committed:
treemaker schema passes tests

lgray committed:
try extended install extras again (except spark)

hmm quoting?

just dask

add parsl

agoose77 commented on coffea/nanoevents/methods/candidate.py:
```diff
@@ -12,14 +12,14 @@
 
 
-@awkward.mixin_class(behavior)
+@awkward.behaviors.mixins.mixin_class(behavior)
```
We expose this under `ak.mixin_class` in the latest v2 pre-releases. I know at one point we did miss this, but it's not necessary to use the long-form any more.

lgray committed:
physlite schema passes tests

lgray committed:
delphes schema passes tests

lgray committed:
repin to spark3.3 (mapInArrow)

lgray committed:
remove python3.8 restrictions on parsl and spark

lgray committed:
bump to java17 for spark

black

lgray committed:
de-lazify Correct*Factory in anticiptation for dask_awkward, skip spark tests while being worked on

lgray committed:
skip tests if no spark

use importerskip to import pyspark

remove old arrow fix

lgray committed:
awkward2 is out of rc

lgray committed:
uproot 5 is out of rc

lgray committed:
do any of parsl/dask/spark work on windows?

interestingly - dask functions on windows

remove spark and parsl which assume linux

add java to windows, remove parsl from macos

figure out spark on windows later - seems doable

comparison ops are hard

lgray committed:
expose awkward forth flag in NanoEventsFactory.from_root

more allowed words

agoose77 commented on coffea/lookup_tools/lookup_base.py:
```diff
@@ -18,33 +18,28 @@ def __call__(self, *args, **kwargs):
             ):
-                nplike = awkward.nplike.of(*inputs)
-                if not isinstance(nplike, awkward.nplike.Numpy):
+                nplike = awkward._nplikes.nplike_of(*args)
```
@lgray we spoke about this on Slack, but in the interest of my not forgetting — on the Awkward side, we'd prefer Coffea not to rely on any private APIs such as `nplikes`. The solution here is to check `ak.backend(*args)`, and exit if it is not equal to the string `"cpu"`.

lgray replied:
yes - have been dealing with other things - was on my list to fix!

agoose77 commented on coffea/lookup_tools/lookup_base.py:
```diff
@@ -18,33 +18,28 @@ def __call__(self, *args, **kwargs):
-        out = awkward._util.broadcast_and_apply(args, getfunction, behavior)
-        assert isinstance(out, tuple) and len(out) == 1
-        return awkward._util.wrap(out[0], behavior=behavior)
+        behavior = awkward._util.behavior_of(*args)
```
`ak.transform` takes care of preserving the behaviour here :)
```suggestion
```

lgray committed:
passes pre-commit

add pre-commit hook instead of old lint

correct naming

remove more old ci

maybe this works

update manifest rules to pass tests

more manifest

more manifest

ignore symlink

lgray committed:
python 3.11 finally?

use miniforge instead of miniconda

perhaps quotes?

lgray committed:
will need to figure out python 3.11 elsewhere

lgray committed:
allow unsecured commands for miniforge

lgray committed:
add isort with black compatibility

lgray committed:
remove errant print from physlite

lgray committed:
pin numpy past security vulnerability

lgray committed:
get local spark tests to work

nsmith- committed:
Skip a step in normalizing buffers for NanoEvents

It seems now all internal layout and index types have a __array__
implementation for np.array() to work. And, the intermediate ak.Array()
  call was triggering from_iter, slowing things down.

nsmith- committed:
Unused import

lgray committed:
dask-awkward is out of alpha

nsmith- committed:
Workaround uproot5 issue with @* leaves, fix TLorentzVector

nsmith- commented on coffea/nanoevents/mapping/uproot.py:
```diff
@@ -47,14 +49,21 @@ def _lazify_form(form, prefix, docstr=None):
-                form["contents"][field], prefix + f",{field},!item"
-            )
+            elif field.startswith("@"):
+                # workaround uproot5 bug
```
Bug is fixed in https://github.com/scikit-hep/uproot5/pull/808

nsmith- committed:
Very sneaky comma

lgray committed:
use dask in factorized jet corrector

lgray committed:
jet resolution is daskified

lgray committed:
all individual correction components daskified

lgray committed:
CorrectedXFactories are daskified but produces really nasty graph

lgray committed:
remove workaround / use ak forth by default

lgray committed:
don't use awkward forth in physlite for now

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
fix not yet released?

lgray committed:
all default ak forth to True

lgray committed:
better implementation of handling dask in lookup_tools, rewrite of dask-aware jet tools

lgray committed:
update dask(-awkward) pins, add dask-histogram, test dak in lookup_tools a bit

lgray committed:
provide more readable names for components of met corrections

lgray committed:
use dictionaries to get return products in test

lgray committed:
add cirrus ci to run all tests on arm

lgray committed:
further dask work

lgray committed:
dask work checkpoint

lgray committed:
remove servicex from awkward2 tests

lgray committed:
get tests going again (reminder to unpin awkward once fix in place

lgray committed:
fix test and unpin awkward2

lgray committed:
fix treemaker tests

lgray committed:
stash work on moving nanoevents to dask

lgray committed:
rebase cruft

lgray committed:
more daskification

lgray committed:
tags for dask-ready schemas

lgray committed:
update isort

lgray committed:
nanoevents dask almost there, TODO: fix weakref pickling hack

lgray committed:
numpy limits for numba

lgray committed:
lookup base, CorrectedXFactory now support necessary columns analysis

agoose77 commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -52,11 +61,22 @@ def getfunction(layout, depth, **kwargs):
+            behavior=item.behavior,
+        )
+    elif backend == "typetracer":
+        zlitem = item.layout.form.length_zero_array()
```
We're trying to avoid the need to use `length_zero_array` in typetracer code unless it's 100% needed. Would you be able to keep tabs on places where you're using `length_zero_array` for workarounds so that we can remove them in future?

lgray replied:
Yes - some of this is going through growing pains as I start using `dask_awkward` as the data source instead of `dak.from_awkward`, the former has rather more intense usage of typetracers.

Once we're on the other side I'll make sure it's minimal and let you know the rest.

lgray replied:
The one you highlight here will probably stay since I need to feed a non-zero-length array to as a PRNG seed.

lgray committed:
let lookup base (dask mode) deal with non-array args to functions

lgray committed:
daskify PackedSelection

lgray committed:
serialization speed improvement for lookups

lgray committed:
we can now use 1.0 instead of faking something in dak

lgray committed:
refactor _ensure_flat; PackedSelection and Weights daskified more clearly

lgray committed:
jetmet tools fully daskified; pins; lookup_base fixes for data-touching

lgray committed:
reset Weights and PackedSelection for redo; start on converting executors to dask

lgray committed:
daskified packed selection in a more correct way

lgray committed:
starting to work on daskifying treemaker

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
get the dask-nanoevents interface looking similar to the old one

lgray committed:
Weights is daskified; various rebase cruft for jetmet tools

lgray committed:
more efficient distribution strategy for heavy lookups, first try

lgray committed:
faster tokenization of corrections in local mode

lgray committed:
treemaker is daskified

lgray committed:
pin to awkward > 2.1.0; switch to calver

lgray committed:
getting closer with PHYSLITE

lgray committed:
delphes is probably daskified

nikoladze committed:
fix form zipping for physlite schema

lgray committed:
Merge pull request #776 from nikoladze/awkward2_dev-fix-physlite-zip

fix form zipping for physlite schema

lgray committed:
Merge branch 'master' into awkward2_dev

lgray committed:
parquet nanoevents + work on physlite

lgray committed:
ressurect parquet nanoevents and associated tests

lgray committed:
now sunder for array context

lgray committed:
xfail tests that need it for now

lgray committed:
2023.4.rc0

lgray committed:
Merge branch 'master' into awkward2_dev

lgray committed:
more xfails, these are heisenbugs...
